Load modules from JSON files

TadeasKucera opened the pull request:
In this PR we start to load modules from JSON files.

TadeasKucera commented on include/yaramod/types/modules/module.h:
```diff
@@ -9,13 +9,14 @@
- * Every module has to provide its features to state what must hold for it to be loaded.
+ * Class representing features of Yaramod.
  */
 enum Features
```
TODO: Move `Features` class to other file. The `Module` class no longer depends on `Features` class

TadeasKucera replied:
Done

TadeasKucera commented:
Currently, the public `.json` files are located in `src/modules/public` and are being transformed into `.cpp` modules in `src/modules`. Do we want to reorganize this somehow?

MatejKastak commented:
Is this file necessary? `src/modules/__init__.py` Since modules in this repository are only related to CPP part of the yaramod, there is no need for this file.

MatejKastak commented:
Nitpick: I have noticed that `*.json` module files are missing a newline at the end of the file. Can you correct that?

MatejKastak commented on src/json_to_byte_array.py:
```diff
@@ -0,0 +1,41 @@
+        if extension == ".json":
+            source_path = os.path.join(input_directory, file)
+            output_file = os.path.join(output_directory, filename + '.cpp')
+            with open(source_path, 'r') as f:
+                module_content = f.read()
+            array = bytearray(module_content.encode('utf-8'))
+
+            with open(output_file, 'wb') as f:
+                print('Created module', output_file, 'from', filename)
+                f.write(array)
```
I don't think this is correct. You are essentially
```
read(file.json) -> convert it to bytearray -> write(file.cpp, bytearray)
```
which is in the end operation that just copies the contents from the json files to cpp files. Bytearray is just a different representation of the same data. What we want is to **preprocess** the json data to a valid C++ language constructs, that can be then used from the C++ yaramod without even accessing the filesystem.

But in your case, if you open generated .cpp files, they are just plain json objects.

However, we have some limitations to keep yaramod portable:
- you must use a construct that allows for "literals" with arbitrary size, as we discussed windows compilers can impose a limitation on a length of a string literal

For example you should be able to use arrays, choosing a proper type is up to you:
```cpp
char MATH_MODULE[] = ['{', '\n', '\t', 'k', 'i', 'n', 'd', ':', ...]
```
then in the `module-pool` you would need to convert it to some type that is compatible with your json library. Maybe using `reinterpret_cast` or some other `*_cast` or doing it by "manually" with C++ function for example. 

After the conversion you should have access to the original `file.json`. I know this might sound complicated but we are doing this to ensure cross platform compatibility and users of our library won't have to do anything to adapt with those changes.

TadeasKucera replied:
I have changed the process similarly as suggested:
The `json_to_array.py` python script goes through `/modules` directory and processes all modules that are present there as `.json` (public modules). Each file is split per char and the sequence of chars is written in a new file `module_<name>_generated.h` that contains the following class:
```
class Gmodule_<name> : public ModuleContent
{
public:
	Gmodule_<name>() : ModuleContent("pe", {'{', '\n', ' ', ' ', ' ', ' ', '\"', 'k', 'i', 'n', 'd', '\"', ':', ' ', '\"', 's', 't', 'r', 'u', 'c', 't',...})
	{
	}
};
```
where `<name>` is shortened name of the input `.json` file and `ModuleContent` is a simple base class prepared to store the sequence. These files are generated in the `/include/yaramod/types/modules/generated` directory together with one additional file `modules_list.h` which includes all generated `.h` modules and creates a vector of `ModuleContent` containing an instance of each of the created modules.

MatejKastak commented:
[Here](https://github.com/avast/yaramod/pull/143/files#diff-ff06f40131180649f3b16e1e8a1ce8008a9f25bfa67830244ea023435063c539R68) you are calling `found_modules = _addDirectory(YARAMOD_PUBLIC_MODULES_DIR);` which should be the directory path of the converted files. This approach would require those files to be present at this path for every machine that installs precompiled yaramod and also you need to guarantee that when installing `pip install yaramod` the modules created during the compilation stay at this location (e.g. `/tmp/pip-req-build-x0p690ty/src/modules`) .

However, when I was playing with your changes this is not true and after the `pip` is done installing `yaramod` it deletes the modules and you are not able to use them anymore.

As we were discussing on the meetings:
1) Adding `custom_command` that converts the json files to cpp representation is correct
2) But you are still reading those files from the filesystem, which is what we want to avoid

MatejKastak reviewed (changes requested):
I think you have done a great deal of work (creating module files, parsing them to internal representation, module-pool, ...) :+1: , however we still need to adjust some things.

My advice would be:
1) Start with the python preprocessor. Think about the representation of the modules in C++. **Resulting files should not be a valid JSON, they should be a valid C++ (that also encodes the content of JSON)**. I have left a comment in the relevant part describing how you can approach this.
2) Create some code that can convert the C++ module representation that you have created in the step before to a representation that is accepted by the JSON library that you added.
3) When you initialize module-pool you should not interact with the filesystem. Work only with the C++ files (**don't read them from the disk, let them be included during the compilation using what is already in the language `#include`**) that you generated from json modules.

I think these are the most important things to get right, after they are resolved we can move to a another iteration with the review.

What do you think? If you have any questions just contact me.

MatejKastak commented:
> Currently, the public .json files are located in src/modules/public and are being transformed into .cpp modules in src/modules. Do we want to reorganize this somehow?
If so do not forget to modify .gitignore

To answer your question, I think that it would be better to keep the out of the `src` directory, as they are not the part of the source code that is compiled, they are used to generate the source code that will be eventually compiled. My suggestion would be to just place them to a folder `GIT_REPO_ROOT/modules`.

Also a suggestion is to create a folder `scripts` for the python script used for preprocessing, as it is not the yaramod source. It's just a tool we use to compile it.

TadeasKucera commented on src/json_to_byte_array.py:
```diff
@@ -0,0 +1,41 @@
+        elif opt in ("-o", "--ofile"):
+            output_directory = arg
+
+    for file in os.listdir(input_directory):
+        filename, extension = os.path.splitext(file)
+        if extension == ".json":
+            source_path = os.path.join(input_directory, file)
+            output_file = os.path.join(output_directory, filename + '.cpp')
+            with open(source_path, 'r') as f:
+                module_content = f.read()
+            array = bytearray(module_content.encode('utf-8'))
+
+            with open(output_file, 'wb') as f:
+                print('Created module', output_file, 'from', filename)
+                f.write(array)
+        else:
+            continue
```
The continue was definitely purposeless. Thank you for your suggestion, I adapted it. (To be honest I don't see much difference :smile: )

TadeasKucera commented on modules_path.h.in:
```diff
@@ -0,0 +1,19 @@
+#define YARAMOD_PUBLIC_MODULES_DIR "${YARAMOD_PUBLIC_MODULES_DIR}"
+#define YARAMOD_PRIVATE_MODULES_DIR "${YARAMOD_PRIVATE_MODULES_DIR}"
+
+#endif /* MODULES_PATH */
```
You are right, thank you! I removed the file and made sure that there are only two places where we work with the file system:
1. First is in the `json_to_array.py` file. In the `json_to_array.py` we generate all the files located in `include/yaramod/types/modules/generated` directory, which are all `module_<name>_generated.h` files containing jsons as an array plus one file `modules_list.h` including all generated modules.
2. Second is in `modules_pool.cpp` using `std::filesystem`, where we access files given by the user by explicitly filling an environmental variable `YARAMOD_MODULE_SPEC_PATH` with the paths of the modules.

TadeasKucera commented:
I agree with moving modules to `/modules` directory and I also moved the python script into a new `/scripts` directory.
Thank you very much for your review and suggestions. I believe the PR is now ready for second review.

MatejKastak commented on src/utils/json.cpp:
```diff
@@ -0,0 +1,63 @@
+/**
+ * @file src/utils/utils.cpp
+ * @brief Implementation of utility functions.
+ * @copyright (c) 2017 Avast Software, licensed under the MIT license
+ */
```
Nitpick: Invalid header

MatejKastak commented on include/yaramod/utils/json.h:
```diff
@@ -0,0 +1,22 @@
+/**
+ * @file include/yaramod/utils/json.h
+ * @brief Declaration of utility functions.
+ * @copyright (c) 2017 Avast Software, licensed under the MIT license
```
Nipick:
```suggestion
 * @copyright (c) 2021 Avast Software, licensed under the MIT license
```

MatejKastak commented on docs/rtd/getting_started.rst:
```diff
@@ -22,18 +22,16 @@ for example need to provide correct types of parameters to module function calls
-world so we improve existing or develop our own modules. What we can we share with the community and in the future, we would really like to see
-all the functions we have also in upstream YARA but that doesn't entirely depend on us. It's not like we are keeping it away becuase we want to keep
-it secret.
+YARA is easily extensible with modules which provide you a way to call C functions from your YARA ruleset. We realize that modules are important
```
Can you also add a section summarizing the different locations that yaramod can consume? I have looked at the code and it seems you can specify the location using `YARAMOD_MODULE_SPEC_PATH` env variable, but I wasn't able to find it here.

What I suggest is creating some kind of short list or hierarchy, that will go through the priority, for example:
1. `YARAMOD_MODULE_SPEC_PATH` env variable
1. Directory passed to yaramod instance constructor - will load custom `.json` files
1. Upstream yara default - builtin to yaramod

Also please state the overwriting policy, as to what happends when you use custom modules. Are other things on the hierarchy ignored?

TadeasKucera replied:
Hello and thank you for all your remarks. I added a section that should clear out both the preference and overwriting.

MatejKastak reviewed (changes requested):
Thanks for the second iteration. Great work :+1:

I have left some comments regarding the symbol documentation not being exposed via python. 

Excuse the time it took me to review this, it is pretty complex to work with.

TadeasKucera commented on scripts/json_to_array.py:
```diff
@@ -0,0 +1,88 @@
+            input_directory = arg
+        elif opt in ("-o", "--ofile"):
+            output_directory = arg
+
```
I somehow forgot to test that, thank you for noticing! :+1:

metthal reviewed (changes requested):
1. There needs to exist way for @MatejKastak to obtain the documentation of individual function overloads (or we can generalize it to documentation of a symbol). This is a crucial part that is missing.
2. I don't like the fact that providing environment variable with paths to modules automatically discards default modules in yaramod. I like the approach outlined by @MatejKastak much more. There should be a priority list and documented what loads first, second etc. I would expect something like

> 1st are default modules, 2nd are modules from specified environment variables and they override default modules, 3rd are modules directly provided to Yaramod constructor and they override everything else

TadeasKucera commented on src/types/modules/module.cpp:
```diff
@@ -67,4 +141,230 @@ bool Module::isInitialized() const
 	return _structure != nullptr;
 }
 
+/**
```
Thank you for pointing this out. The Symbol class has now a read-only property `documentation` and the usage is shown in the last two representation tests.

TadeasKucera commented:
Thank you @metthal for your notes. 
1. I incorporated the documentation and names of function parameters into Yaramod and exposed it so we can access the data from Python.
2. I modified this behaviour by working with two env variables `YARAMOD_MODULE_SPEC_PATH` and `YARAMOD_MODULE_SPEC_PATH_EXCLUSIVE`. The first one adds supplied modules to the default modules (or the modules supplied to `Yaramod` constructor). The second one makes the modules replace the default modules.

I also modified the Wiki to tell how the modules can be added / replaced. It is now ready for next review.

MatejKastak reviewed (approved):
LGTM

Thanks for your work. :+1: You have done a great refactor and added good ways to make 
yaramod more flexible.
